src/python/pants/engine/rules.py

```diff
@@ -439,6 +440,19 @@ def dependency_optionables(self):
     """A tuple of Optionable classes that are known to be necessary to run this rule."""
     return ()
 
+  @classmethod
+  def _validate_type_field(cls, type_obj, description):
+    if not isinstance(type_obj, type):
+      raise TypeError(f'{description} provided to @rules must be types! Was: {type_obj}.')
+    if dataclasses.is_dataclass(type_obj):
+      if not (type_obj.__dataclass_params__.frozen or
+              getattr(type_obj, frozen_after_init.sentinel_attr, False)):
+        raise TypeError(
+          f'{description} {type_obj} is a dataclass declared without `frozen=True`, or without '
+          'both `unsafe_hash=True` and the `@frozen_after_init` decorator! '
+          'The engine requires that fields in params are immutable for stable hashing!')
+    return type_obj
```

Comment:
Could we easily extend this to other banned types? I'm thinking `List`, `Set`, and `FrozenSet` for now.

Reply:
See https://github.com/pantsbuild/pants/pull/8548#discussion_r350046300 -- I think that hashability might be a sufficient sanity check for the types we're concerned with here?